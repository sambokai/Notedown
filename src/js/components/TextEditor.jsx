import React from 'react';
import PropTypes from 'prop-types';

class TextEditor extends React.Component {
  handleChange = (e) => {
    this.props.onUpdateNote(e.target.value);
  };

  render() {
    const textAreaStyle = {
      resize: 'none',
    };

    return (
      <textarea
        className="w-100"
        rows={15}
        style={textAreaStyle}
        placeholder={
          (this.props.note.id === -1) ? this.props.noNotesMessage : this.props.placeholder
        }
        disabled={(this.props.note.id === -1)}
        value={this.props.note.body}
        onChange={this.handleChange}
        ref={this.props.textAreaRef}
      />
    );
  }
}

TextEditor.propTypes = {
  placeholder: PropTypes.string,
  noNotesMessage: PropTypes.string,
  note: PropTypes.shape({
    body: PropTypes.string,
    id: PropTypes.number,
  }).isRequired,
  onUpdateNote: PropTypes.func.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  textAreaRef: PropTypes.object.isRequired,
};

TextEditor.defaultProps = {
  placeholder: 'Write your notes in here...',
  noNotesMessage: "Create a new note by clicking the 'New Note' Button",
};

export default React.forwardRef((props, ref) =>
  <TextEditor {...props} textAreaRef={ref} />);
