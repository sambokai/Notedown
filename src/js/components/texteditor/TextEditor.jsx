import React from 'react';
import PropTypes from 'prop-types';

class TextEditor extends React.Component {
  constructor(props) {
    super(props);

    this.textArea = React.createRef();
  }

  componentDidMount() {
    this.focusTextArea();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.note.id !== this.props.note.id) {
      this.focusTextArea();
    }
  }

  focusTextArea() {
    if (this.textArea.current) {
      this.textArea.current.focus();
    }
  }

  handleChange = (e) => {
    this.props.onUpdateNote(e.target.value);
  };

  emptyNotePassed() {
    return Object.keys(this.props.note).length === 0;
  }


  render() {
    const textAreaStyle = {
      resize: 'none',
    };

    return (
      <textarea
        id="text-editor-textarea"
        className="w-100 h-100 mousetrap"
        rows={15}
        style={textAreaStyle}
        placeholder={
          (this.props.note.id) ? this.props.placeholder : this.props.noNoteMessage
        }
        disabled={this.emptyNotePassed()}
        value={this.emptyNotePassed() ? '' : this.props.note.body}
        onChange={this.handleChange}
        ref={this.textArea}
      />
    );
  }
}

TextEditor.propTypes = {
  placeholder: PropTypes.string,
  noNoteMessage: PropTypes.string,
  note: PropTypes.shape({
    body: PropTypes.string,
    id: PropTypes.number,
  }).isRequired,
  onUpdateNote: PropTypes.func.isRequired,
};

TextEditor.defaultProps = {
  placeholder: 'Write your notes in here...',
  noNoteMessage: 'No note selected',
};

export default TextEditor;
