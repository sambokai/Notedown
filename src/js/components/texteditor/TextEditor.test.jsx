import React from 'react';

import { render } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';

import TextEditor from './TextEditor';

const onUpdateNoteMock = jest.fn();

describe('<TextEditor/>', () => {
  it('should render correctly if a note is passed', () => {
    const wrapper = render((
      <TextEditor
        note={{ body: 'This is a note.', id: 12412 }}
        onUpdateNote={onUpdateNoteMock}
      />
    ));

    expect(renderToJson(wrapper)).toMatchSnapshot();
  });

  it('should render a disabled html-textarea if NO note is passed', () => {
    const wrapper = render((
      <TextEditor
        note={{}}
        onUpdateNote={onUpdateNoteMock}
      />
    ));

    expect(renderToJson(wrapper)).toMatchSnapshot();
  });
});
