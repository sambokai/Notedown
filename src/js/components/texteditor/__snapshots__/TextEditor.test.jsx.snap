// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextEditor/> should render a disabled html-textarea if NO note is passed 1`] = `
<textarea
  class="w-100 h-100 mousetrap"
  disabled=""
  placeholder="No note selected"
  rows="15"
  style="resize:none"
/>
`;

exports[`<TextEditor/> should render correctly if a note is passed 1`] = `
<textarea
  class="w-100 h-100 mousetrap"
  placeholder="Write your notes in here..."
  rows="15"
  style="resize:none"
>
  This is a note.
</textarea>
`;
